### PR TITLE
Shifted the name of the Necromancy skill in the secondary skill bar l…

### PR DESCRIPTION
**Pull request by JasonQorbin**
This PR addresses issue #7491. The change makes th UI render more like the original game by rendering the name of the Necromancy skill one pixel to the left of center so that the long skill name doesn't overlap with the gold border on the righthand side of the secondary skill tile. The solution is not perfect because the text has a shadow that now touches the left side border but the original game looked like this also.

A few things to note about the code changes:

1. I opted to add a check to see if the current skill is the necromancy skill and handle the special case instead of rounding the result of the padding calucation down because I didn't want to affect all the exiting content (which no one has complained about and is probably happy with).
2. The problem pointed out by @ihhub is still relevant: This only fixes the visual bug for the english version of the game and for long strings in other languages we will need a better solution. Handling a special case for every long skill name in every language is not feasible.
3. I added a lambda for the centering logic which is repeated several times in this function. This may be beyond the scope of the issue. Happy to revert that back.

Happy to take guidance on this to arrive at a better more general solution

### CODE REVIEW

**Comment by JasonQorbin** [comment]
Sorry. Updated formatting with clang-format.

**Comment by ihhub** [comment]
Hi @JasonQorbin , the root of the problem is in presence of shadow within the font which is 1 pixel wide. The shadow is present at left and bottom sides. Therefore, in order to tackle the problem we need to get width of the text, width of the icon and calculate with 1 pixel shift:
```cpp
xPos = ( iconWidth - textWidth + 1 ) / 2;
```
Essentially, the whole fix is
```cpp
text.draw( pos.x + ( pos.width - text.width() + 1 ) / 2, pos.y + 53, dstsf );
```

Please give a try.

**Comment by JasonQorbin** [comment]
I didn't realize I could squish the text. Thanks for that. I'll take another look this evening.

**Comment by JasonQorbin** [comment]
Hi @ihhub. I've tried a few approaches.

The fix you recomended solves the problem of the name overlapping the righthand border, which was the original problem we were trying to solve. However, it still casts a shadow on the lefthand border like so:

Necromancy with 1px adjustment

I then attempted the suggestion from @Branikolog in the original issue of "cutting the shadow for such cases".
So in the current version of the PR, if the text is just wide enough that the shadow would fall on the border then
I repaint that section of the border to remove the shadow. This is the result:

Necromancy with 1px adjustment and border repainted

**Comment by ihhub** [comment]
Hi @JasonQorbin , I am adding changes to limit rendering of text within a given ROI. See #8101. With this approach it will be easier to avoid any hacky code.

**Comment by JasonQorbin** [comment]
Text, centered, no more shadow, and no more hacky redrawing of the border thanks to the new text drawing function.
Screenshot 2023-12-10 204357

File: src/fheroes2/gui/skill_bar.cpp [hunk 1]
```diff
@@ -238,14 +238,15 @@ void SecondarySkillsBar::RedrawItem( Skill::Secondary & skill, const fheroes2::R
 
     if ( use_mini_sprite ) {
         const fheroes2::Text text{ std::to_string( skill.Level() ), fheroes2::FontType::smallWhite() };
-        text.draw( pos.x + ( pos.width - text.width() ) - 3, pos.y + pos.height - 10, dstsf );
+        text.drawInRoi( pos.x + ( pos.width - text.width() ) - 3, pos.y + pos.height - 10, dstsf, pos );
     }
     else {
         fheroes2::Text text{ Skill::Secondary::String( skill.Skill() ), fheroes2::FontType::smallWhite() };
-        text.draw( pos.x + ( pos.width - text.width() ) / 2, pos.y + 5, dstsf );
+        int skillNamePaddingX = ( pos.width - text.width() ) / 2 - 1;
```
Comment (ihhub) [inline src/fheroes2/gui/skill_bar.cpp:245]: This variable can be const.

**Review by ihhub — COMMENTED** [review]
Now the code is very nice and neat :)

**Comment by ihhub** [comment]
Hi @Branikolog , please test these changes since you requested them.

File: src/fheroes2/gui/skill_bar.cpp (hunk 1 above)
Comment (JasonQorbin) [reply src/fheroes2/gui/skill_bar.cpp]: Done. Also mirrored the same variable for the skill level text for consistency.

**Review by ihhub — APPROVED** [review]

**Comment by Branikolog** [comment]
Hello, @JasonQorbin !
Thanks for working on this issue.
Necromancy skill name aligned properly in the hero info window now. 👍 
But lvlup window still has the issue.
image

**Comment by ihhub** [comment]
Hi @Branikolog , the level up window is a separate issue.

**Comment by ihhub** [comment]
Hi @Branikolog , I suggest to check other languages and how it looks for skill names.

**Comment by Branikolog** [comment]
Hi, @ihhub 
Okay.
What about skill info window?

image

It is aligned properly, but there's a shadow on a frame, while in hero info window shadow was dismissed for the frame?

**Comment by Branikolog** [comment]
@ihhub 

>Hi @Branikolog , I suggest to check other languages and how it looks for skill names.

Thanks to the ability to select language in the adventure map settings it was really easy and comfortable. :)

Symbols are cut for the names on other languages:

image

image

Such implementation is not OK in my opinion.

**Comment by JasonQorbin** [comment]
I'm happy to apply the fix the other menus and windows. But I would appreciate some guidance on what sort of intervention would be considered OK for handling languages that, it would appear, the UI wasn't designed for. In that last screenshot, the text of the skill levels of adjacent skill would overlap if they aren't truncated. And those are single words so we can't render it out on two separate lines.

Maybe we can consider scaling the text down? Not sure that would look good and very long words would become illegible very quickly at some resolutions.

Maybe combined with scaling the UI up? Might be fine on higher resolutions and larger screens but that wouldn't work at the resolution shown in the screenshots above.

I'm just kicking around ideas to try to spark something. I'm not really sure what the answer is either.

**Comment by JasonQorbin** [comment]
Hi everyone.

I've changed my local copy to truncate the text with ellipses ("...") if it is any longer than the max width of the block in the Hero Screen as suggested by @zenseii. 

Excuse the weird character encoding, but I think it gets the point across.

Screenshot 2023-12-14 001606

@Branikolog if you're happy with this then I'll update the pull request with this change and then we can create a new issue(s) to fix both the level-up window and the one that pops up when you right-click the skill.

**Comment by Branikolog** [comment]
Hi, @JasonQorbin and @zenseii 
Yeah, that's a nice solution. We have some other names too long for these skill boxes. So showing "..." is definitely better than text overlapping the nearest icon text.
image

**Comment by oleg-derevenetz** [comment]
Hi @JasonQorbin there is something weird about this PR, too many changed lines (4036 lines added, 2276 lines removed, 70 files affected). Please sync it with `master` properly (e.g. by merging `master` into this PR's branch).

**Comment by JasonQorbin** [comment]
Apologies. Should be fixed now. Should only be 11 lines in one file.

I rebased my commits together and ended up overwriting other peoples commits. Terribly sorry!!

**Review by Branikolog — APPROVED** [review]

File: src/fheroes2/gui/skill_bar.cpp [hunk 2]
```diff
@@ -237,15 +237,23 @@ void SecondarySkillsBar::RedrawItem( Skill::Secondary & skill, const fheroes2::R
...
+        fheroes2::Text text{ std::to_string( skill.Level() ), fheroes2::FontType::smallWhite() };
+        if ( text.width() > 76 )
+            text.fitToOneRow( pos.width );
+        text.drawInRoi( pos.x + ( pos.width - text.width() ) - 3, pos.y + pos.height - 10, dstsf, pos );
     }
     else {
         fheroes2::Text text{ Skill::Secondary::String( skill.Skill() ), fheroes2::FontType::smallWhite() };
-        text.draw( pos.x + ( pos.width - text.width() ) / 2, pos.y + 5, dstsf );
+        if ( text.width() > 76 )
+            text.fitToOneRow( pos.width );
```
Comment (ihhub) [inline src/fheroes2/gui/skill_bar.cpp:248]: Please add parenthesis for this if-body:
```cpp
if ( text.width() > 76 ) {
    text.fitToOneRow( pos.width );
}
```

The same applies for the code below.

File: src/fheroes2/gui/skill_bar.cpp [hunk 3]
```diff
@@ -237,15 +237,23 @@ void SecondarySkillsBar::RedrawItem( Skill::Secondary & skill, const fheroes2::R
     fheroes2::Blit( sprite, dstsf, pos.x + ( pos.width - sprite.width() ) / 2, pos.y + ( pos.height - sprite.height() ) / 2 );
 
     if ( use_mini_sprite ) {
-        const fheroes2::Text text{ std::to_string( skill.Level() ), fheroes2::FontType::smallWhite() };
-        text.draw( pos.x + ( pos.width - text.width() ) - 3, pos.y + pos.height - 10, dstsf );
+        fheroes2::Text text{ std::to_string( skill.Level() ), fheroes2::FontType::smallWhite() };
+        if ( text.width() > 76 )
+            text.fitToOneRow( pos.width );
+        text.drawInRoi( pos.x + ( pos.width - text.width() ) - 3, pos.y + pos.height - 10, dstsf, pos );
```
Comment (ihhub) [inline src/fheroes2/gui/skill_bar.cpp:243]: I don't think that we need to modify this code. There is literally only 1 digit to render. I suggest to revert these changes. This part of code is executed in Kingdom Overview dialog.

File: src/fheroes2/gui/skill_bar.cpp [hunk 4]
```diff
@@ -237,15 +237,23 @@ void SecondarySkillsBar::RedrawItem( Skill::Secondary & skill, const fheroes2::R
...
     }
     else {
         fheroes2::Text text{ Skill::Secondary::String( skill.Skill() ), fheroes2::FontType::smallWhite() };
-        text.draw( pos.x + ( pos.width - text.width() ) / 2, pos.y + 5, dstsf );
+        if ( text.width() > 76 )
```
Comment (ihhub) [inline src/fheroes2/gui/skill_bar.cpp:247]: We must avoid hardcoded values. This condition should be based on **sprite** width which is 75 pixels. In other words, the condition should be `if ( text.width() > sprite.width() + 1 )`.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @JasonQorbin , I left several comments in this pull request. Can you please take a look a them?

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@JasonQorbin , thank you so much for your contribution and patience with this issue!

**Pull request merged**